### PR TITLE
feat: make VM reboot timeout configurable via --vm-reboot-timeout

### PR DESCRIPTION
**Pull request by ovi-ionos**
### Description of your changes

This PR adds:

- a `--vm-reboot-timeout` flag to be able to configure the amount of time waited after a VM has been rebooted when using the `statemap` feature
- a boolean flag `--extend-serverset-timeout-for-vm-reboot` to optionally be able to extend the timeout of the **`ServerSet` controller**, without needing to change the global timeout _for all other controllers_. This is necessary, otherwise a `--vm-reboot-timeout` longer than the global `--timeout` would not actually be applied... This is a compromise, to avoid having to raise `--timeout` for all controllers to unnecessarily high values.
- a check to exit early with an error at startup when `--vm-reboot-timeout` is configured manually beyond the `--timeout` without also setting `--extend-serverset-timeout-for-vm-reboot`. This is intended to guard against unintentional misconfiguration, and will **only** fire if the user _manually specifies the `--vm-reboot-timeout` flag_, and never with the default configuration (which already has --timeout lower than --vm-reboot-timeout). One could instead log a warning, but I think it's likely that would not be noticed and hence not stop accidental misconfiguration...
- a fix for `WaitForResource` in 3c814471d74ca76931490cd3aba5b1bffc900b22. The calculated deadline wasn't actually properly passed to the function callback, it always received the overall outer context without the more specific deadline. This should not matter, because the callbacks usually complete fast.

Instead of this (a bit over-specific) `--extend-serverset-timeout-for-vm-reboot` flag, one could also introduce a `--timeout-per-resource` flag mirroring `--max-reconcile-rate-per-resource`. That would make the timeout for any controller in this provider configurable, and then one just has to know to increase the timeout for SSet.
If you'd prefer that solution, let me know.

## Checklist

I have:

- [x] Add PR name as appropriate (e.g. `feat`/`fix`/`doc`/`test`/`refactor`)
- [x] Run `make reviewable` and `make crds.clean` to ensure the PR is ready for review
- [x] Add or update tests (if applicable)
- [ ] Add or update Documentation using `make docs.update` (if applicable)
- [x] Update `docs/CHANGELOG.md` file (label: `upcoming release`)
- [x] Check Sonar Cloud Scan

### CODE REVIEW

File: internal/controller/serverset/setup.go [hunk 1]
```diff
@@ -32,6 +32,10 @@ func SetupServerSet(mgr ctrl.Manager, opts *utils.ConfigurationOptions) error {
 		kube: mgr.GetClient(),
 		log:  logger,
 	}
+	reconcileTimeout := opts.GetTimeout()
+	if opts.GetExtendServerSetTimeoutForVMReboot() {
+		reconcileTimeout = max(reconcileTimeout, opts.GetVMRebootTimeout())
+	}
```
[the remark this answers is not in the document]
Comment (ovi-ionos) [reply internal/controller/serverset/setup.go:38]: Addressed with better documentation of the flag.

File: cmd/provider/main.go [hunk 2]
```diff
@@ -51,19 +51,21 @@ import (
...
+		maxReconcileRate                  = app.Flag("max-reconcile-rate", "The global maximum rate per second at which resources may checked for drift from the desired state.").Default("10").Int()
+		timeout                           = app.Flag("timeout", "Timeout duration cumulatively for all the calls happening in the reconciliation functions.").Default("1h").Duration()
+		vmRebootTimeout                   = app.Flag("vm-reboot-timeout", "Timeout to wait for a VM to reboot and report as ready again after a failover, for ServerSets using a custom state map. Must not exceed --timeout, or the wait will be cut short by the overall reconcile deadline, unless --extend-serverset-timeout-for-vm-reboot is set.").Default("120m").Duration()
+		extendServerSetTimeoutForVMReboot = app.Flag("extend-serverset-timeout-for-vm-reboot", "If true, widens the ServerSet controller's own reconcile timeout to be at least --vm-reboot-timeout, so a long VM reboot wait is not cut short by --timeout. Only affects ServerSet reconciles; every other resource kind keeps using --timeout unchanged.").Default("false").Bool()
```
[the remark this answers is not in the document]
Comment (ovi-ionos) [reply cmd/provider/main.go]: Flag documentation improved, and added fail-fast check against misconfiguration.

File: internal/utils/configuration.go [hunk 3]
```diff
@@ -52,6 +56,22 @@ func (o *ConfigurationOptions) GetTimeout() time.Duration {
 	return o.Timeout
 }
 
+// GetVMRebootTimeout returns the value for the VMRebootTimeout option
+func (o *ConfigurationOptions) GetVMRebootTimeout() time.Duration {
+	if o == nil {
+		return 0
+	}
```
[the remark this answers is not in the document]
Comment (ovi-ionos) [reply internal/utils/configuration.go:83]: I don't think it's worth adding tests for this here, none of the other timeout flags are tested like this.

I added tests for `isFlagSetByUser`.

File: cmd/provider/main.go [hunk 4]
```diff
@@ -51,19 +51,22 @@ import (
...
+		leaderElection                    = app.Flag("leader-election", "Use leader election for the controller manager.").Short('l').Default("false").Envar("LEADER_ELECTION").Bool()
+		createGracePeriod                 = app.Flag("create-grace-period", "Grace period for creation of IONOS Cloud resources.").Default("1m").Duration()
+		maxReconcileRate                  = app.Flag("max-reconcile-rate", "The global maximum rate per second at which resources may checked for drift from the desired state.").Default("10").Int()
+		timeout                           = app.Flag("timeout", "Timeout duration cumulatively for all the calls happening in the reconciliation functions.").Default("1h").Duration()
+		vmRebootTimeoutFlag               = app.Flag("vm-reboot-timeout", "Timeout to wait for a VM to reboot and report as ready again after a failover, for ServerSets using a custom state map. This budget is shared across every replica that needs to reboot within the same reconcile call (e.g. an image update touching all replicas) - it is not a fresh window per replica. If set higher than --timeout, --extend-serverset-timeout-for-vm-reboot must also be set, or the provider will refuse to start.")
```
[the remark this answers is not in the document]
Comment (ovi-ionos) [reply cmd/provider/main.go:63]: Correct. Each replica gets its own window, but the _whole loop shares one reconcile deadline_ from the controller timeout. So the doc is correct.

File: internal/controller/serverset/serverset.go [hunk 5]
```diff
@@ -553,7 +560,7 @@ func (e *external) updateServersFromTemplate(ctx context.Context, cr *v1alpha1.S
 				e.log.Info("Server has been updated and uses custom state map, waiting for reboot to finish", "serverset", cr.Name, "server", servers[idx].Name)
 				// After reboot, wait for server to be in running state again before proceeding to next replica
 				if err := kube.WaitForResource(
-					ctx, kube.VMRebootTimeout, func(ctx context.Context, mapName, mapNamespace string) (bool, error) {
+					ctx, e.vmRebootTimeout, func(ctx context.Context, mapName, mapNamespace string) (bool, error) {
 						return e.isVMSoftwareRunning(ctx, requestTimestamp, servers[idx].Name, mapName, mapNamespace)
 					}, cr.Spec.ForProvider.Template.Spec.StateMap.Name, cr.Spec.ForProvider.Template.Spec.StateMap.Namespace,
 				); err != nil {
```
[the remark this answers is not in the document]
Comment (ovi-ionos) [reply internal/controller/serverset/serverset.go:574]: Ok, this is a pre-existing bug actually. Fixed in 3c814471d74ca76931490cd3aba5b1bffc900b22.

I don't think it really has an impact, because "Kubernetes API calls inside isVMSoftwareRunning can outlive" the total timeout, is a problem of a matter of _milliseconds_ normally. Kubernetes API calls complete fast usually.

**Review by cristiGuranIonos — DISMISSED** [review]
might be a little better  to put a per-resource timeout, but this should be fine since it will only ever be set in these conditions

**Review by cristiGuranIonos — APPROVED** [review]
